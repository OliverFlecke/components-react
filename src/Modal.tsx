import React, { FC, useEffect, useMemo } from 'react';
import ReactDOM from 'react-dom';

export interface ModalProps {
	children: any;
	isOpen: boolean;
	onDismiss?: () => void;
}

const Modal: FC<ModalProps> = props => {
	if (!props.isOpen) return null;

	return <Container {...props} />;
};

export default Modal;

const Container: FC<ModalProps> = ({ children, onDismiss }) => {
	const { node } = useNodeInRoot('modal-root', {
		className: 'absolute top-0 left-0 w-screen h-screen',
	});

	const render = (
		<div
			className="h-full max-w-full center bg-opacity-50 bg-white dark:bg-opacity-50 dark:bg-black"
			onClick={onDismiss}
		>
			<div
				className="bg-white dark:bg-black rounded-md m-4"
				onClick={e => e.stopPropagation()}
			>
				{children}
			</div>
		</div>
	);

	return ReactDOM.createPortal(render, node);
};

function getOrCreateRoot(id: string): HTMLElement {
	let root = document.getElementById(id);

	if (!root) {
		const body = document.getElementsByTagName('body')[0];
		root = document.createElement('div');
		root.id = id;
		body.appendChild(root);
	}

	return root;
}

function useNodeInRoot(id: string, options?: unknown) {
	const node = useMemo(() => document.createElement('div'), []);
	if (options) {
		Object.assign(node, options);
	}

	useEffect(() => {
		const root = getOrCreateRoot(id);
		root.appendChild(node);

		return () => {
			root?.removeChild(node);
		};
	}, [node, id]);

	return { node };
}
