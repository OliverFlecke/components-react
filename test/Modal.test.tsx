import React from 'react';
import Modal from '../src/Modal';
import renderer from 'react-test-renderer';

describe('Modal', () => {
	test.each([[true], [false]])('Render snapshot', open => {
		const tree = renderer
			.create(
				<Modal isOpen={open} onDismiss={() => console.log('dismiss')}>
					<div></div>
				</Modal>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
