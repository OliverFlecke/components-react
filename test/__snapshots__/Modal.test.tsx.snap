// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Render snapshot 1`] = `
<div
  className="absolute top-0 left-0 w-screen h-screen"
>
  <div
    className="h-full max-w-full center bg-opacity-50 bg-white dark:bg-opacity-50 dark:bg-black"
    onClick={[Function]}
  >
    <div
      className="bg-white dark:bg-black rounded-md m-4"
      onClick={[Function]}
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Modal Render snapshot 2`] = `null`;
